### DISCUSSION TITLE
Is there a technical reason why there are only callbacks for batch `complete`, `success`, and `death`?

### DISCUSSION THREAD

**Discussion by dhughes** [sidekiq/sidekiq#5836] [Q&A]
I have a complicated system I've created for a client that syncs eCommerce data from one system to another. I can get into details about this system, but this is it in a nutshell:

Two jobs run in a batch. When they're done, a `complete` callback triggers a third job to run. If either of (or both of) the first two jobs don't succeed on the first run, Sidekiq still fires the `complete` callback as soon as both jobs have been tried once. This means that job 3 can be started before the first two have really finished (successfully or unsuccessfully). 

The problem I have is that most retries in my system are due to recoverable problems. For example, SSL handshake errors, intermittent flakiness in the two systems I'm syncing data between, cosmic rays, etc. And, the two initial jobs are relatively long running. They enqueue a bunch of other smaller jobs, but have to get data from system 1 via its API, one page at a time, within rate limits. So, if either one of these encounters a hiccup, the `complete` callback is fired, causing my third job to run, even though jobs 1 and 2 aren't really done yet and are likely to succeed. 

Side note: jobs 1 and 2 can potentially enqueue a ton of other smaller jobs within a child batch. Some of these jobs may fail and be retried until death. Most of these retries will succeed, but it's possible that some may die.

Anyhow, I don't think any of the three callbacks works for me:

* `complete` - fires once each job has been tried once. This is the scenario I'm in right now. I can't depend on job 1 or 2 really being _done_ when this callback is fired. If a job failed, it can be retried and might ultimately succeed or die. Thus, job 3 can run before I'd like it to.
* `success` - This is wonderful, if all jobs before job 3 succeed. But what if I have a job in my batch that legitimately can't succeed? Like, maybe one of the smaller jobs has bad data and can't write to system 2? In that case, this callback is never fired. It does seem appropriate at this callback never happens, since not all of the jobs succeeded.
* `death` - This is fired the first time a job completely dies. That doesn't feel like it really applies to my scenario. I mean, it's possible that I might have 500 jobs that are retrying due to a network hiccup. If one of those jobs exhausts its retries, that doesn't mean that the other 499 will also die. It just means that this one job has died.

In the perfect world, I’d wish for a finish callback that that only fires after all of the jobs in a batch have finished successfully or died after exhausting all of their retries. With that hook I could guarantee that I’ve made my best effort to run all of the small jobs that spawn from the first two jobs, before kicking off the third job.

I realize that Sidekiq is the way that it is because Mike has done the hard work and made the tough calls. I know batches are complicated, especially as it relates to weird edge cases around errors. I'm not really asking for any changes here.

In the end, I'm wondering if there is a technical reason that allows for the everything-has-run-once `complete` callback, the everything-has-succeeded `success` callback, and the a-single-job-has-died `death` callback, but not a callback to indicate that all jobs have either succeeded or died. I'd love to learn more about that.

**Comment by mperham** [comment]
> * `success` - This is wonderful, if all jobs before job 3 succeed. But what if I have a job in my batch that legitimately can't succeed? Like, maybe one of the smaller jobs has bad data and can't write to system 2? In that case, this callback is never fired. It does seem appropriate at this callback never happens, since not all of the jobs succeeded.
> * `death` - This is fired the first time a job completely dies. That doesn't feel like it really applies to my scenario. I mean, it's possible that I might have 500 jobs that are retrying due to a network hiccup. If one of those jobs exhausts its retries, that doesn't mean that the other 499 will also die. It just means that this one job has died.

But it also means that the batch will **never** succeed. The idea is that **death** and **success** are mutually exclusive. Either all jobs succeed or at least one of the jobs will die.

> 
> In the perfect world, I’d wish for a finish callback that that only fires after all of the jobs in a batch have finished successfully or died after exhausting all of their retries. With that hook I could guarantee that I’ve made my best effort to run all of the small jobs that spawn from the first two jobs, before kicking off the third job.

Right, maybe you can mock this :finished callback by adding both a :death and :success callback. You should get one or the other.

> In the end, I'm wondering if there is a technical reason that allows for the everything-has-run-once `complete` callback, the everything-has-succeeded `success` callback, and the a-single-job-has-died `death` callback, but not a callback to indicate that all jobs have either succeeded or died. I'd love to learn more about that.

Well, the system doesn't track outstanding dead batch jobs today so I have no way of knowing whether all outstanding batch jobs have reached a terminal state. In other words, Sidekiq doesn't know if any pending jobs are still retrying or if all have died. We just know when the first pending job dies but other jobs in the batch could still retry and finish successfully.

It sounds like maybe that's the crux of the issue: the :death callback is fired upon the first job death but it sounds like you want the :death callback to also mean that no more jobs for the batch will execute after the death callback? I don't think that's a guarantee I can make. wdyt?

**Reply by dhughes** [reply]
> The system doesn't track outstanding dead batch jobs today.

Ah. That sounds like the root of my issue. If a batch job is dead, then I can understand why you wouldn't want to keep track of it. It's not going to do anything else and just takes up space, right? Thus, as soon as any one batch job dies, then it's gone, and so long.

That does lead to a followup question. How do you know when all jobs have succeeded? I'd imagine you wouldn't keep data about these job around after the fact either.

I was just looking at the `Sidekiq::Batch::Info`. I see that it has integer properties for `completed`, `failures`, `pending`, and `total`. I wonder if you could add another for `dead`? And maybe when `completed + dead` equals `total` then fire an event? I know essentially nothing about Redis, so I'm not sure if this is possible or practical. 

Another idea might be to have a callback that fires each time any job in the batch runs. With that, I could look at the batch status' `pending` value. If that's zero then I think that would mean that all jobs have either succeeded or died, right? I suppose a downside to this might be handling a callback a zillion times for large batches.

Maybe one other option would be to fire a callback each time a job in the batch dies, not just the first time one of them dies. I could follow the same strategy, check the status to see if there are any more pending jobs. If I use both this hypothetical `job_death` callback and the `success` callback, I could have all paths covered. 

I'm curious to know if callbacks are expensive? Would it be possible to add a bunch of callbacks and just let engineers decide which ones to listen to? I'm mostly thinking of callbacks for individual jobs, but they could be for anything in the job and batch lifecycle. Off the top of my head I can think of `job_added_to_batch`, `job_started`, `job_finished`, `job_succeeded`, `job_failed`, `job_died`, `batch_created`, `batch_expired`, `batch_became_immutable`, `batch_canceled`,  etc. I can think of use cases for some, but not most of these. But I'm just one guy.

Failing any other options, I guess I could just go with the least bad option, which is what you suggested, just use both `death` and `success` callbacks.

**Reply by mperham** [reply]
> How do you know when all jobs have succeeded?

Every Batch has a counter of failed jobs and pending jobs. Complete fires when failed == pending. Success fires when pending == 0.

> I'm curious to know if callbacks are expensive?

In this case they are separate jobs so they are far more expensive than a method invocation. Besides, I dislike the Rails pattern of stuffing callbacks absolutely everywhere. Not going to happen.

The Batch API is meant to provide a simple(?) facade on top of a really hard problem: coordinating and monitoring a distributed workload. The code is really racy and extremely difficult to get correct. Changing callbacks might seem like a small change but could actually be hard or impossible to implement.

If you want to suggest changes, the best way is by documenting a good usecase that can't be easily implemented today with the Batch API. Don't look inside the batch code and suggest tweaks to the implementation.

**Reply by dhughes** [reply]
> Besides, I dislike the Rails pattern of stuffing callbacks absolutely everywhere.

You make a compelling point. 🤣 

I'm guessing you don't consider the scenario I outlined above to be compelling?

**Reply by mperham** [reply]
It sounds like you want a `finished` callback which fires as the last action for a batch. If we have counts for pending, failed and dead, we may be able to determine when to fire `finished`.

**Reply by dhughes** [reply]
I realize this is _way_ late of a reply, but I just saw your reply on March 27. I would be _really_ happy if there were a `finished` callback added. That would be terrific!

**Reply by hadaromash** [reply]
Funny thing, I'm also working on a syncing eCommerce data from my app to another and I just came up with a verysimilar issue and found this thread while trying to come up with the solution. This `finished` callback is exactly what I need to solve my use case as well.

**Comment by mperham** [comment]
I know. I have yet to determine if I can implement it in a safe and reliable manner.

**Reply by dhughes** [reply]
That's quite fair. I have my fingers crossed for the future.

**Comment by dhughes** [comment]
I just wanted to add one other point on this thread. I've been trying to switch over to using `on_success` instead of `on_complete`, but have found that the batch is immutable, so I can't add more jobs to the batch by the time the `on_success` callback has run. So that means I really do need to use `on_complete`, which has the limitation that some workers may still be pending a retry.
